NixOS 25.05 allows max 100 users

I'm upgrading nixos 24.05 to 25.05 and the rebuild fails if I have more than 100 users on the server:

```
...
/nix/store/7sb1nkpf82nb5kj7qc4bbqkwj1l1mdv9-update-users-groups.pl: out of free UIDs or GIDs
Activation script snippet 'users' failed (2)
...
Failed to run activate script
reloading user units for root...
restarting sysinit-reactivation.target
the following new units were started: set-disk-quotas.service
warning: error(s) occurred while switching to the new configuration
Error: Process completed with exit code 1.
```

If I'm not mistaken, this change https://github.com/NixOS/nixpkgs/commit/46600598f1344d4c5029a883103b0b4bd1728235#diff-d019ca3e4041b6a9883c95ca3098db7477255070ad92cdb6d2c3ef7a04b38773R338 made it so that by using autoSubUidGidRange we only can have up to 100 users. If we have more `allocId` will raise `out of free UIDs or GIDs`. This error is a bit misleading because the problem is subuid/subgid ranges.

I think we need to change it:

from `my ($min, $max, $delta) = (100000, 100000 + 100 * 65536, 65536);` to  `my ($min, $max, $delta) = (100000, 100000 + 10000 * 65536, 65536);`.

After further investigation I realised that even if I don't have +100 users on a server I could enounter this issue. If I had 100 users in any time on my nixos server I get this error because the historic data is stored in `/var/lib/nixos/auto-subuid-map`.

To see how many users we have we can run `jq 'length' /var/lib/nixos/auto-subuid-map`. If you have more than 100 you need to `mv /var/lib/nixos/auto-subuid-map /var/lib/nixos/auto-subuid-map.bak`. On the next nixos rebuild it will be recreated with the current users.

I also had to define subUidRanges and subGidRanges for every user. This can become a bit tricky if you have a lot of them or they are semi-dynamically created. This code works for me (I have 100 users):

```
...
  indexOf = x: xs:
    let
      len = builtins.length xs;
      indices = builtins.genList (i: i) len;
      matches = builtins.filter (i: builtins.elemAt xs i == x) indices;
    in
      if matches == [] then null else builtins.head matches;
  count = 10000;

  # attribute set
  users = builtins.listToAttrs (builtins.map (dbUser: {
    name = dbUser.user;
    value = {
      ...
    };
    # we don't need subuid ranges for now because we don't use docker/podman
    # but we need to define them to avoid error when running nixos rebuild
    # "update-users-groups.pl: out of free UIDs or GIDs"
    autoSubUidGidRange = false;
    subUidRanges = [{startUid = 100000 + (indexOf dbUser allUsers) * count; count = count;}];
    subGidRanges = [{startUid = 100000 + (indexOf dbUser allUsers) * count; count = count;}];
  }) allUsers);
...
```

--- a/nixos/modules/config/update-users-groups.pl
+++ b/nixos/modules/config/update-users-groups.pl
@@ -334,8 +334,13 @@ $subUidsPrevUsed{$_} = 1 foreach values %{$subUidMap};
 sub allocSubUid {
     my ($name, @rest) = @_;
 
-    # TODO: No upper bounds?
-    my ($min, $max, $delta) = (100000, 100000 + 100 * 65536, 65536);
+    # The upper bound of 29000 users is derived from limits in
+    # nixos/modules/programs/shadow.nix which allocates the UID ranges
+    # 1000-29999 for regular users. System users are not allocated
+    # subordinate user or group IDs, and so after removing those 999
+    # system users, we end up with 29000 users that are non-system users
+    # that need subUIDs and subGIDs.
+    my ($min, $max, $delta) = (100000, 100000 + 29000 * 65536 - 1, 65536);
     my $prevId = $subUidMap->{$name};
     if (defined $prevId && !defined $subUidsUsed{$prevId}) {
         $subUidsUsed{$prevId} = 1;

